src/main/java/org/candlepin/model/PoolCurator.java

```diff
@@ -1159,13 +1256,45 @@ public List<Pool> getSubPoolsForStackIds(Collection<Consumer> consumers, Collect
         return output;
     }
 
-    @SuppressWarnings("unchecked")
+    /**
+     * Retrieves pools that have a {@SourceSubscription} that has a source stack ID equal to the provided
+     * stack ID.
+     *
+     * @param owner
+     *  the owner of the pools to retrieve
+     *
+     * @param stackId
+     *  the ID of the stack for the source subscriptions to retrieve pools for
+     *
+     * @return the pools that have a source subscription with a source stack ID equal to the provided
+     *  stack ID.
+     */
     public List<Pool> getOwnerSubPoolsForStackId(Owner owner, String stackId) {
-        return createSecureCriteria()
-            .createAlias("sourceStack", "ss")
-            .add(Restrictions.eq("ss.sourceStackId", stackId))
-            .add(Restrictions.eq("owner", owner))
-            .list();
+        if (owner == null || stackId == null || stackId.isBlank()) {
+            return new ArrayList<>();
+        }
+
+        EntityManager entityManager = this.getEntityManager();
+        CriteriaBuilder builder = entityManager.getCriteriaBuilder();
+        CriteriaQuery<Pool> query = builder.createQuery(Pool.class);
+        Root<Pool> root = query.from(Pool.class);
+
+        Join<Pool, SourceStack> stackJoin = root.join(Pool_.sourceStack);
+
+        List<Predicate> predicates = new ArrayList<>();
+        Predicate securityPredicate = this.getSecurityPredicate(Pool.class, builder, root);
+        if (securityPredicate != null) {
+            predicates.add(securityPredicate);
+        }
+
+        predicates.add(builder.equal(root.get(Pool_.owner), owner));
+        predicates.add(builder.equal(stackJoin.get(SourceStack_.sourceStackId), stackId));
+
+        query.select(root)
+            .where(predicates.toArray(new Predicate[0]));
```

Comment:
There's a lot of instances where `.toArray(new Predicate[0])` is called, which is a bit confusing. Especially the 0-length array part, since the array being passed in is supposed to be what will be used to store the list items (so how is a 0-length array supposed to store anything?). Looking a bit into it, it turns that toArray will check the size of the array argument, and if its not high enough to fit the list items, it throws it away and creates a new array with appropriate size, which is basically exactly the same as what `.toArray()` (without an argument) does.

So switching all `.toArray(new Predicate[0])` into `toArray()` still works and would read better

Reply:
Unfortunately, I don't think that this will work. `List.toArray()` returns `Object[]`. You can cast `Object[]` to `Predicate[]` and that will work at compile time, but you will get a `ClassCastException` at runtime.

Reply:
ah, didn't notice the different return type. I guess the current way will have to do

Reply:
As a bit of an FYI: there are some arcane bits here happening under the hood of `.toArray` and its clunky interface. The methods predate generics or collection typing, which is why the API is so weird and verbose. A modern take on them would probably do something like we've seen (and implemented) in other areas where we declare a "type" object that captures the destination class and then it returns a properly typed array from that. But, as with many things surrounding the collections interface and generics, bad decisions made in the transition to 1.5 kinda hamstring it a bit even today.

In terms of performance, the 0-length array call is actually our best option due to it bypassing length checks and Java-land initialization. That is, when we allocate an "empty" array like this, we don't need to do anything other than create the ref and typing information, nor do we have to zero out an array we plan to completely fill anyway. The underlying workhorse logic which does the reallocation and copying *also* skips the length check and zeroing/nulling out of the new array, which means we save O(n) amount of work that we don't care about anyway.